Boost python gen tool

lukedirtwalker opened the pull request.

lukedirtwalker committed:
Initial commit of PythonWrapperGenerator

lukedirtwalker committed:
Mostly working version of PythonWrapperGenerator

Still needs some fixes:
-Only Wrap classes which have a certain baseclass (e.g. Node)
-Correctly wrap attributes with const & return value
-What to do about begin-/endModification?
-Wrap functions
-Detect enum Attributes (e.g. methodKind)

lukedirtwalker committed:
Wrap ATTRIBUTE_VALUE attributes as well

lukedirtwalker committed:
Wrap implicit attributes like methodKind

But at the same time make sure that private attributes are not exposed.

lukedirtwalker committed:
Correctly wrap methods which return a const reference

lukedirtwalker committed:
Wrap ATTRIBUTE_OOP_ANNOTATIONS macro

lukedirtwalker committed:
Correctly handle function which return a pointer

Also make sure that implicit attributes are handled the same way as
normal attributes

lukedirtwalker committed:
Add a workaround wrapper

lukedirtwalker committed:
Use clang's getQualifiedName instead of doing it manually

This fixes a problem for the StatementItemList where the base was
previously exported as Model::TypedList<StatementItem>, now the
StatementItem template argument is with the correct namespace prefix.

lukedirtwalker committed:
Wrap lines which are longer than a certain length (120)

lukedirtwalker committed:
Remove blacklist hack instead have a baseclass whitelist

We only want to wrap classes which have a certain base class, e.g.
Model::Node for now.

lukedirtwalker committed:
Print the wrappers in the correct order

Base classes have to be wrapped before its subclasses.

lukedirtwalker committed:
Make the tool work on ModelBase as well

lukedirtwalker committed:
Don't wrap private stuff

lukedirtwalker committed:
Use clang qualified name even more

lukedirtwalker committed:
Remove header file check

It didn't work reliably. For example VariableDeclaration wrappers were
not generated.

lukedirtwalker committed:
Handle abstract classes better

lukedirtwalker committed:
Wrap all the various TypedList's

Inspired from: https://wiki.python.org/moin/boost.python/StlContainers

For now we just offer iteration and length methods, we can always extend
this.

lukedirtwalker committed:
Remove space before newline

lukedirtwalker committed:
Implement config and clean some stuff up

lukedirtwalker committed:
Print license on top of generated file

lukedirtwalker committed:
Workaround StatementItemList that is currently not working

lukedirtwalker committed:
Merge remote-tracking branch 'generator/master' into boost_python_gen

lukedirtwalker committed:
Clean stuff after merging PythonWrapperGenerator

Add project to Envision's subprojects

lukedirtwalker committed:
Use relative path for PythonWrapperGenerator config

dimitar-asenov commented on PythonWrapperGenerator/main.cpp:
```diff
@@ -0,0 +1,18 @@
+	GenTool t;
+	t.setSubDirPath(Config::instance().envisionReadPath());
+	t.run();
+	qDebug() << "Done!";
```
Don't use `qDebug` for actual output of the program. Isn't this the reason you included `iostream`?

dimitar-asenov commented on PythonWrapperGenerator/llvm-config-envision.sh:
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
```
This file looks identical to the one in `CppImport`. We should absolutely not have two files. We should put a single file in `../misc` where there are a few other scripts, and use it from both projects.

lukedirtwalker replied:
True it is a copy of it, I will move it

dimitar-asenov commented on PythonWrapperGenerator/TypeUtilities.cpp:
```diff
@@ -0,0 +1,157 @@
+				auto argument = templateSpecialization->getArg(0);
+				return templateArgToString(argument);
+			}
+			else qDebug() << "Not 1 argument in super template";
```
I see there are still quite a few uses of `qDebug`. Is the idea to remove them later as the code matures?

dimitar-asenov commented on PythonWrapperGenerator/README.md:
```diff
@@ -0,0 +1,5 @@
+#BoostPythonGen
```
I suppose we no longer need this, since we'll merge the code in Envision's tree

dimitar-asenov commented on PythonWrapperGenerator/PythonWrapperGenerator.pro:
```diff
@@ -0,0 +1,62 @@
+TEMPLATE = app
+CONFIG += console
+CONFIG -= app_bundle
+
```
I mentioned this in a previous commit, but I'm not sure if it will be shown properly, so here it is again:
- we should use a precompiled header with all the stl+clang libraries

dimitar-asenov commented:
Alright, I had a brief look, but I feel overwhelmed by this stuff. We'll need to get together and look at it.

lukedirtwalker committed:
PythonWrapperTool: move code to /src and use precompiled header

The source code move is to make it more consistent with the rest of
Envision and using a precompiled header should improve compile speed.

lukedirtwalker committed:
Move llvm-config-envision.sh to common directory

No need to keep a copy in 2 different directories

lukedirtwalker committed:
Remove PythonWrapperGenerator README.md

lukedirtwalker committed:
Remove qDebug usage and clean code a bit

Use std::cout for application output.
Use Q_ASSERT(false) for missing implementation.

lukedirtwalker committed:
Make APIData a singleton

We only need one instance during the whole execution of the
PythonWrapperGenerator so it makes sense.

lukedirtwalker committed:
Use RAII for APIPrinter

This is much nicer than to manually open / close the out file.

lukedirtwalker commented:
Okay I cleaned this up a bit more. There is still a TODO and a FIXME which we currently don't matter that much. So I think this can be merged once you feel is a good moment to do it.

dimitar-asenov commented:
Thanks, Lukas.
